server/aesdsocket.c

```diff
@@ -372,181 +396,370 @@ int run_daemon(void)
- *
- * @return Returns 0 on successful operation, -1 otherwise.
+ * This function does the following:
+ * 1. Receives data from the client and writes it to the file /var/tmp/aesdsocketdata
+ * 2. Reads the content from the same file and sends it back to the client.
+ * 
+ * @param thread_param Pointer to the thread data structure
+ * @return Returns the pointer to the thread data structure
  */
-int receive_and_store_data(void)
+void* client_data_handler(void *thread_param)
 {
-    ssize_t bytes_received = 1;              ///< Number of bytes received
-    char buf[BUF_LEN];                       ///< Temporary buffer to hold received data
-    int ret_status;                          ///< Status returned by the write operation
-    size_t total_bytes_received = 0;         ///< Total bytes received so far
-    char* dynamic_buf = NULL;                ///< Dynamic buffer to accumulate received data
+    int result;
+    // variables for receiving data
+    ssize_t bytes_received = 0;
+    char receive_buffer[BUF_LEN];
+
+    // variables for sending data
+    ssize_t bytes_sent = 0;
+    char send_buffer[BUF_LEN];
+    ssize_t bytes_from_file = 1;
+
+    memset(receive_buffer, 0, BUF_LEN);
+    memset(send_buffer, 0, BUF_LEN);
 
-    // Loop to receive and accumulate data
-    while (bytes_received > 0)
+    ClientThreadData_t *thread_data_ptr = (ClientThreadData_t*)thread_param;
+
+    inet_ntop(thread_data_ptr->pClientAddr->ss_family,
+              get_in_addr((struct sockaddr *)&(thread_data_ptr->pClientAddr)),
+              s, sizeof s);
+    syslog(LOG_INFO, "New connection established: %s", s);
+
+    syslog(LOG_INFO, "Thread %ld initialized", thread_data_ptr->threadId);
+
+    // Lock mutex to protect file
+    result = pthread_mutex_lock(thread_data_ptr->pMutex);
+    if (result == ERROR)
     {
-        memset(&buf, 0, sizeof(buf));        // Clear the temporary buffer
+        syslog(LOG_ERR, "Failed to acquire mutex");
+        return NULL;
+    }
 
-        bytes_received = recv(clientSocketFd, buf, BUF_LEN, 0);
+    // Initialize the condition variable
+    void *newline_found = NULL;
 
-        // Check for receive errors
+    // Execute loop as long as the condition variable is NULL
+    while (newline_found == NULL)
+    {
+        bytes_received = recv(thread_data_ptr->clientSocketFd, receive_buffer, BUF_LEN, 0);
         if (bytes_received == ERROR)
         {
-            syslog(LOG_ERR, "Receive failed");
-            s_flags.command_status_success = false;
-            free(dynamic_buf);  // Free allocated memory before returning
-            return -1;
+            syslog(LOG_ERR, "Data reception unsuccessful");
+            return NULL;
         }
 
-        // Reallocate memory for dynamic buffer
-        char* new_dynamic_buf = realloc(dynamic_buf, total_bytes_received + bytes_received);
-        if (new_dynamic_buf == NULL)
+        result = write(dataFileDescriptor, receive_buffer, bytes_received);
```

Comment:
@Suraj-Ajjampur what would happen if a packet was longer than BUF_LEN with this design?
Seems like your earlier design for A5 was efficient. Why change it?